capstone_project/src/test/java/com/google/sps/ResourceTest.java

```diff
@@ -0,0 +1,102 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.sps.data.Resource;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Class that tests the ability to add Resource Entities to Datastore */
+@RunWith(JUnit4.class)
+public final class ResourceTest {
+  private static DatastoreService datastore;
+  private String PREEXISTING_CATEGORY = "Depression";
+  // number of the preexisting resources in addPreexistingResources()
+  private final int NUM_PREEXISTING = 4;
+
+  private LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  @Before
+  public void setUp() {
+    helper.setUp();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void checkAddPreexistingResourcesToEmptyDatastore() {
+    // First check that there are no Resources in Datastore
+    int expected = 0;
+    assertTrue(equalsDatastoreCount(expected));
+
+    // call addPreexistingResources(), which adds 4 hardcoded Resources to Datastore
+    Resource.addPreexistingResources();
+
+    expected = NUM_PREEXISTING;
+    assertTrue(equalsDatastoreCount(expected));
+  }
+
+  @Test
+  public void checkAddPreexistingResourcesNoDuplicates() {
+    // First add a Resource Entity to Datastore that shares the same
+    // category as a preexisting Resource found in Resource.java, then call
+    // addPreexistingResources(). If the duplicate is added, the count would be greater
+    // than the number of preexisting Resources
+    Resource.storeResource(PREEXISTING_CATEGORY, "");
+    assertTrue(equalsDatastoreCount(1));
+
+    Resource.addPreexistingResources();
```

Comment:
Maybe break this up into 2 tests so each one gets its own assertTrue(), one test for adding the single preexisting resource and one for adding all the 4 preexisting resources?

Reply:
I think datastore gets reinitialized on every test. So, there wouldn't be a way to test for duplicates if it was split into two. I think what Lynda is doing right now is: Add a Category that preexisting method will add, then call the preexisting method and ensure it was not added twice. I agree that there kind of is a bit too much asserting for one test. I think this could just be narrowed down to: Asserting the count of PREEXISTING_CATEGORY to be one. The other asserts don't seem necessary imo, in regards to the test name

Reply:
Datastore does get reinitialized, so I can't split the test. +1 to Leo's explanation. I've edited so that only the last assert is kept!